sklearn/discriminant_analysis.py

```diff
@@ -464,64 +472,67 @@ def _solve_svd(self, X, y):
         y : array-like of shape (n_samples,) or (n_samples, n_targets)
             Target values.
         """
+        xp, is_array_api = get_namespace(X)
+
+        if is_array_api:
+            svd = xp.linalg.svd
+        else:
+            svd = scipy.linalg.svd
+
         n_samples, n_features = X.shape
-        n_classes = len(self.classes_)
+        n_classes = self.classes_.shape[0]
 
         self.means_ = _class_means(X, y)
         if self.store_covariance:
             self.covariance_ = _class_cov(X, y, self.priors_)
 
         Xc = []
         for idx, group in enumerate(self.classes_):
-            Xg = X[y == group, :]
-            Xc.append(Xg - self.means_[idx])
+            Xg = X[y == group]
+            Xc.append(Xg - self.means_[idx, :])
 
-        self.xbar_ = np.dot(self.priors_, self.means_)
+        self.xbar_ = self.priors_ @ self.means_
 
-        Xc = np.concatenate(Xc, axis=0)
+        Xc = xp.concat(Xc, axis=0)
 
         # 1) within (univariate) scaling by with classes std-dev
-        std = Xc.std(axis=0)
+        std = xp.std(Xc, axis=0)
         # avoid division by zero in normalization
         std[std == 0] = 1.0
         fac = 1.0 / (n_samples - n_classes)
 
         # 2) Within variance scaling
-        X = np.sqrt(fac) * (Xc / std)
+        X = math.sqrt(fac) * (Xc / std)
```

Comment:
are they strictly equivalent? Shouldn't this be `xp.sqrt` instead?

Reply:
Looking at this again, I think it's better to do `xp.asarray()` on the scalar, so we can use `xp.sqrt` on it.

Reply:
lolol, why? I was convinced with you last comment.

Reply:
For me, the pros and cons of `math` vs `xp.asarray` on Python scalars is very balanced. The argument for using `xp.asarray` is that it forces us to be in "array land" and not need to think about any Python scalar + Array interactions.

Although, the [ArrayAPI spec](https://data-apis.org/array-api/latest/API_specification/type_promotion.html#mixing-arrays-with-python-scalars) does state that that `python_scalar * array` is the same as `xp.asarray(python_scalar, dtype=dtype_of_array) * array`.

REF: https://discuss.scientific-python.org/t/poll-future-numpy-behavior-when-mixing-arrays-numpy-scalars-and-python-scalars/202